apps/sms/js/drafts.js

```diff
@@ -32,183 +26,107 @@
-     *
-     * @param  {Draft} draft draft to delete.
-     *
-     * @return {Drafts} return the Drafts object.
+     * Deletes a draft record from Drafts in-memory collection.
+     * @param  {Draft} draft Draft instance to delete.
+     * @returns {Drafts} Returns the Drafts object.
      */
     delete: function(draft) {
-      var thread;
-      var index;
-
-      if (draft) {
-        thread = draftIndex.get(draft.threadId);
-        if (!thread) {
-          return this;
-        }
-        index = thread.indexOf(draft);
+      if (!draft || !(draft instanceof Draft)) {
+        throw new Error('Draft should be valid Draft instance.');
+      }
 
-        // For cases where the provided "draft" object
-        // could not be found by object _identity_:
-        //  - If a thread was found by draft.threadId, but
-        //    draft had no id property, delete all drafts
-        //    for this threadId
-        //  - Otherwise, the draft object might be a copy,
-        //    or manually composed "draft object", so iterate
-        //    the drafts and look for the one matching the
-        //    provided draft.id.
-        if (index === -1) {
-          if (thread && typeof draft.id === 'undefined') {
-            thread.length = 0;
-          } else {
-            if (thread) {
-              thread.some(function(stored, i) {
-                if (stored.id === draft.id) {
-                  index = i;
-                  return true;
-                }
-              });
-            }
-          }
-        }
+      var drafts = draftIndex.get(draft.threadId);
+      if (!drafts || !drafts.length) {
+        return this;
+      }
 
-        if (index > -1) {
-          thread.splice(index, 1);
+      // For drafts bound to thread we should remove all drafts we have, but for
+      // thread-less drafts we should remove draft with the same id only.
+      if (draft.threadId) {
+        drafts.length = 0;
+      } else {
+        var indexToRemove = drafts.findIndex(
+          (storedDraft) => storedDraft.id === draft.id
+        );
+
+        if (indexToRemove > -1) {
+          drafts.splice(indexToRemove, 1);
         }
       }
+
       return this;
     },
+
     /**
-     * byThreadId
-     *
-     * Returns all the drafts for the specified thread id.
-     *
-     * Calling with `null` will return a `Draft.List` object
-     * containing all of the threadless draft objects.
-     *
-     * eg.
-     *
-     *   Drafts.byThreadId(null)
-     *
-     *
-     *
-     * @param  {Number}  id thread id of the drafts to return.
-     *
-     * @return {Draft.List}  return Draft.List containing drafts for thread id.
+     * Returns draft for the specified thread id or null if it doesn't exist.
+     * @param {Number} threadId Thread id to recall draft for.
+     * @returns {Draft}
      */
-    byThreadId: function(id) {
-      return new Drafts.List(draftIndex.get(id));
+    byThreadId: function(threadId) {
```

Comment:
Yeah, it sounds good to me! The only thing that still bothers me is that ```byThreadId``` and ```byDraftId``` applied to different non-intersected sets of drafts. Or maybe it's just me :)